### PR TITLE
Fix issue with pullrequests in Azure DevOps

### LINKED ISSUE #3180: [Bug] Don't inherit semver on an Azure DevOps PullRequest
**Describe the bug**
When creating a PullRequest on an Azure DevOps repository, no matter you autoincrement minor or patch based on your gitversion.xml configuration, it doesn't inherit from source branch on PullRequest completion. 

## Expected Behavior

It should works as in local, it inherits semver when you merge source branch to target branch configured to inherit. Although it doesn't work on local when you tried to simulate Azure DevOps environment.

## Actual Behavior
When you merge source branch to target branch configured to inherit, it works on a merge but not in a PullRequest in Azure DevOps. Also it doesn't work reproducing an Azure DevOps environment locally.

## Possible Fix
After debugged it, it only requires to detect PullRequest properly on a ReferenceName class.

## Steps to Reproduce
GitVersion.Tool 5.10.3 installed on Azure DevOps.

image

Content of gitversion.xml

```
assembly-versioning-scheme: MajorMinorPatch
assembly-file-versioning-scheme: MajorMinorPatch
assembly-informational-format: '{InformationalVersion}'
mode: ContinuousDelivery
branches: 
  main:
    regex: ^main$
    tag: rc
    increment: None
    is-source-branch-for:
    - bug
    - feature
    prevent-increment-of-merged-branch-version: true
    track-merge-target: false
    tracks-release-branches: false
    is-release-branch: false
    is-mainline: true
    pre-release-weight: 55000
  release:
    regex: ^releases[/-]
    tag: ''
    increment: Inherit
    source-branches:
    - main
    is-source-branch-for:
    - retrofit-bug
    - retrofit-feature
    prevent-increment-of-merged-branch-version: false
    track-merge-target: false
    tracks-release-branches: false
    is-release-branch: true
    is-mainline: true
    pre-release-weight: 55000
  pull-request:
    regex: ^pull[/-]
    tag: 'beta.{BranchName}'
    increment: Inherit
    source-branches:
    - retrofit-bug
    - retrofit-feature
    - bug
    - feature
    prevent-increment-of-merged-branch-version: false
    track-merge-target: false
    tracks-release-branches: false
    is-release-branch: false
    is-mainline: false
    pre-release-weight: 30000
  retrofit-bug:
    regex: ^retrofit/bugs[/-]
    tag: beta.{BranchName}
    increment: Patch
    source-branches: 
    - release
    prevent-increment-of-merged-branch-version: false
    track-merge-target: true
    tracks-release-branches: false
    is-release-branch: false
    is-mainline: false
    pre-release-weight: 30000
  retrofit-feature:
    regex: ^retrofit[/-]
    tag: beta.{BranchName}
    increment: Minor
    source-branches:
    - release
    prevent-increment-of-merged-branch-version: false
    track-merge-target: true
    tracks-release-branches: false
    is-release-branch: false
    is-mainline: false
    pre-release-weight: 30000
  bug:
    regex: ^bugs[/-]
    tag: beta.{BranchName}
    increment: None
    source-branches:
    - main
    prevent-increment-of-merged-branch-version: false
    track-merge-target: false
    tracks-release-branches: false
    is-release-branch: false
    is-mainline: false
    pre-release-weight: 30000
  feature:
    regex: ^(?!.*main)(?!.*releases)(?!.*pull)(?!.*retrofit)(?!.*bugs).*
    tag: beta.{BranchName}
    increment: None
    source-branches:
    - main
    prevent-increment-of-merged-branch-version: false
    track-merge-target: false
    tracks-release-branches: false
    is-release-branch: false
    is-mainline: false
    pre-release-weight: 30000
ignore:
  sha: []
merge-message-formats: {}
```

Results of a PullRequest completion that doesn't inherit incremented Patch.

image

Results of a PullRequest with the fix.

image

## Context
Without this fix, we can not use GitVersion in Azure DevOps.

### PR DESCRIPTION
**Pull request by enriqueraso**
Fix issue with PullRequests in Azure DevOps. PullRequest not well detected and Remotes not well replaced.

## Description
It detects correctly when a reference name is a PullRequest and also remove Remotes properly on a PullRequest.
Helps on Azure DevOps PullRequest completion. Otherwise it doesn't inherit semver calculated on source branch.

## Related Issue
Fix #3180

## Motivation and Context
I want to use GitVersion in Azure DevOps with the gitversion,xml described in the bug.

## How Has This Been Tested?
I tested it on Azure DevOps adding binaries to the repo and change yml file.

## Screenshots:

image

image

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by arturcic** [comment]
@enriqueraso can you update the tests to cover the additional prefixes you added? https://github.com/GitTools/GitVersion/blob/808f8ba06ecc2d7e0b2abdf27e691a7eeaeb5a91/src/GitVersion.App.Tests/PullRequestInBuildAgentTest.cs#L11

**Comment by enriqueraso** [comment]
@arturcic Indeed, I will update it. :)

**Comment by enriqueraso** [comment]
@arturcic additional prefixes were added as well a new unit test.

**Comment by arturcic** [comment]
@enriqueraso, I think if you want to have this included in the next release, better rebase this PR on [support/5.x](https://github.com/GitTools/GitVersion/tree/support/5.x) branch. Please include only your commits on top

File: src/GitVersion.Core/Git/ReferenceName.cs [hunk 1]
```diff
@@ -75,9 +80,17 @@ private string RemoveRemote()
     {
         var isRemote = IsPrefixedBy(Canonical, RemoteTrackingBranchPrefix);
 
-        return isRemote
-            ? Friendly.Substring(Friendly.IndexOf("/", StringComparison.Ordinal) + 1)
-            : Friendly;
+        if (isRemote)
+        {
+            var isPullRequest = IsPrefixedBy(Canonical, RemotePullRequestPrefix1)
+                || IsPrefixedBy(Canonical, RemotePullRequestPrefix2);
```
Comment (asbjornu) [inline src/GitVersion.Core/Git/ReferenceName.cs:86]: Can you perhaps extract `IsPullRequest` to a property so we get a little less repetition among line 28-31 and 85-86?

File: src/GitVersion.Core/Git/ReferenceName.cs [hunk 2]
```diff
@@ -13,6 +13,8 @@ public class ReferenceName : IEquatable<ReferenceName?>, IComparable<ReferenceNa
     private const string TagPrefix = "refs/tags/";
     private const string PullRequestPrefix1 = "refs/pull/";
     private const string PullRequestPrefix2 = "refs/pull-requests/";
+    private const string RemotePullRequestPrefix1 = "refs/remotes/pull/";
+    private const string RemotePullRequestPrefix2 = "refs/remotes/pull-requests/";
```
Comment (asbjornu) [inline src/GitVersion.Core/Git/ReferenceName.cs:17]: If these prefixes are usually checked together, perhaps we can group them into a single array instead?

**Review by asbjornu — CHANGES_REQUESTED** [review]
Please re-target and rebase on `support/5.x`. Besides a few comments, I think this looks good!

File: src/GitVersion.Core/Git/ReferenceName.cs (hunk 1 above)
Comment (enriqueraso) [reply src/GitVersion.Core/Git/ReferenceName.cs:86]: `IsPullRequest` was already a property, so I refactored a little bit the class.

File: src/GitVersion.Core/Git/ReferenceName.cs (hunk 2 above)
Comment (enriqueraso) [reply src/GitVersion.Core/Git/ReferenceName.cs:17]: Moved to `PullRequestPrefixes` array and also added a method `IsPrefixedBy`.

**Review by arturcic — APPROVED** [review]

**Comment by enriqueraso** [comment]
@asbjornu, @arturcic any date for the new NuGet package from support/5.x?

**Pull request merged**